### PR TITLE
Code Insights: Implement new "Add or remove insights" view 

### PR DESCRIPTION
**Pull request by vovakulikov**
Closes https://github.com/sourcegraph/sourcegraph/issues/46291

This PR adds new add or remove insights UI. The new UI has 
- **New visual styles**. We use more available space on the screen that show more data at once. It became more closer to other similar views (like Fuzzy Finder UI)
- Fully keyboard accessible UI. Prior to this PR the old view didn't have any keyboard navigation short cuts. Thanks to https://github.com/sourcegraph/sourcegraph/pull/46362 MultiCombobox UI we do have now a full fledged keyboard navigation. 
- **Performance improvements** First of all there is no loading state anymore. We read all needed data from the cache without unnecessary requests at the beginning. The new view supports smart pagination + search over insights by its title and series label.

| Before | After |
| ---------- | --------- |
| Screenshot 2023-01-16 at 18 37 32 | Screenshot 2023-01-16 at 18 37 12 |

## Test plan
- Go to the dashboard and click add or remove insights button 
- Add and remove some insights with new UI
- Click save and check that all insights that you remove have been removed and all that you added have been added to the dashboard.

## App preview:

- [Web](https://sg-web-vk-add-new-add-insight-ui.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (0.00 kb) | 0.06% (+8.62 kb) | 0.07% (+8.62 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/vk/add-new-add-insight-ui/compare-report.html;tab=live_object) for a full comparison between the commits 48de6bc and 8ffcb9b or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

File: CHANGELOG.md [hunk 1]
```diff
@@ -35,6 +35,7 @@ All notable changes to Sourcegraph are documented in this file.
 ### Changed
 
 - Code Insights no longer uses a custom index of commits to compress historical backfill and instead queries the repository log directly. This allows the compression algorithm to span any arbitrary time frame, and should improve the reliability of the compression in general. [#45644](https://github.com/sourcegraph/sourcegraph/pull/45644)
+- Code Insights has got a new UI for the "add or remove insights" view [#46538](https://github.com/sourcegraph/sourcegraph/pull/46538)
```
Comment (leonore) [inline CHANGELOG.md:38]:
```suggestion
- Code Insights has a new UI for the "Add or remove insights" view, which also allows you to search code insights by series label. [#46538](https://github.com/sourcegraph/sourcegraph/pull/46538)
```
Github doesn't allow me to do this but I think you should add this entry at the bottom of the list

File: enterprise/internal/insights/resolvers/insight_view_resolvers.go [hunk 2]
```diff
@@ -1187,6 +1187,19 @@ func (r *InsightViewQueryConnectionResolver) computeViews(ctx context.Context) (
 			args.UniqueID = unique
 		}
 
+		if r.args.ExcludeIds != nil {
```
Comment (leonore) [inline enterprise/internal/insights/resolvers/insight_view_resolvers.go:1190]: I think you want to merge this with main right, you made a separate PR with this?

File: client/web/src/enterprise/insights/core/backend/gql-backend/deserialization/create-insight-view.ts [hunk 3]
```diff
@@ -42,9 +41,7 @@ export const createInsightView = (insight: InsightViewNode): Insight => {
             const { appliedFilters } = insight
             // We do not support different time scope for different series at the moment
             const step = getDurationFromStep(insight.dataSeriesDefinitions[0].timeScope)
-            const repositories = uniq(
-                insight.dataSeriesDefinitions.flatMap(series => series.repositoryScope.repositories)
-            )
+            const { repositories } = getInsightRepositories(insight.repositoryDefinition)
```
Comment (leonore) [inline client/web/src/enterprise/insights/core/backend/gql-backend/deserialization/create-insight-view.ts:44]: for any curious passerby, this made me check whether we allow duplicates in the repo UI, which we do, but we must deduplicate somewhere because we still only count results for 1 repos (https://sourcegraph.sourcegraph.com/insights/aW5zaWdodF92aWV3OiIyS1MwRHgxaFJQVVVpWlV5dlh0R2h3WERYSjAi)

**Review by leonore — COMMENTED** [review]
just passing by as a code insights contributor- this looks great!

File: enterprise/internal/insights/resolvers/insight_view_resolvers.go (hunk 2 above)
Comment (vovakulikov) [reply enterprise/internal/insights/resolvers/insight_view_resolvers.go]: I think there was conflict between main and this branch, I rebased main and this change has gone away.

File: client/web/src/enterprise/insights/core/backend/gql-backend/deserialization/create-insight-view.ts (hunk 3 above)
Comment (vovakulikov) [reply client/web/src/enterprise/insights/core/backend/gql-backend/deserialization/create-insight-view.ts:44]: @leonore prior to this PR we used deprecated repositoryScope query in each data series definition, but since all series share the same set of repositories we moved this query on the top level and there is no more need to do a uniq filtering here. 

If customers somehow end up with repositories duplicates I think we should dedupe them on the backend

File: client/web/src/enterprise/insights/pages/dashboards/dashboard-view/components/add-insight-modal/AddInsightModal.module.scss [hunk 4]
```diff
@@ -1,12 +1,110 @@
 .modal {
-    padding: 0.5rem 1rem 1rem 1.5rem;
     display: flex;
     flex-direction: column;
-    width: 100%;
-    max-width: 45rem;
+    height: 80vh;
+    max-height: 850px;
+    width: 80vw !important;
```
Comment (valerybugakov) [inline client/web/src/enterprise/insights/pages/dashboards/dashboard-view/components/add-insight-modal/AddInsightModal.module.scss:6]: Is this related to CSS order in production issue?

File: client/web/src/enterprise/insights/pages/dashboards/dashboard-view/components/add-insight-modal/AddInsightModal.tsx [hunk 5]
```diff
@@ -30,79 +35,184 @@ export interface AddInsightModalProps {
...
+                    Add insight to <q>{dashboard.title}</q>
+                </H2>
+
+                <Button variant="icon" className={styles.closeButton} aria-label="Close" onClick={onClose}>
+                    <Icon aria-hidden={true} svgPath={mdiClose} />
+                </Button>
+            </header>
+
+            {/* eslint-disable-next-line react/forbid-elements */}
+            <form className={styles.form} onSubmit={handleSubmit}>
```
Comment (valerybugakov) [inline client/web/src/enterprise/insights/pages/dashboards/dashboard-view/components/add-insight-modal/AddInsightModal.tsx:86]: Curious, if we still need this eslint rule at this point?

File: client/web/src/enterprise/insights/pages/dashboards/dashboard-view/components/add-insight-modal/AddInsightModal.tsx [hunk 6]
```diff
@@ -30,79 +35,184 @@ export interface AddInsightModalProps {
...
+                        {connection && (
+                            <span>
+                                {plural('result', connection.nodes.length)}
+                                {connection.totalCount && <> out of {plural('total', connection.totalCount)}</>}
+                            </span>
+                        )}
+                    </small>
+
+                    <MultiComboboxList
+                        items={connection?.nodes ?? []}
```
Comment (valerybugakov) [inline client/web/src/enterprise/insights/pages/dashboards/dashboard-view/components/add-insight-modal/AddInsightModal.tsx:118]: Does that mean that the `items` array is updated on every render, which causes `setSuggestOptions(items)` inside of the `MultiComboboxList`, which might cause a lot of rerenders since it updates the top level object? (might not a be a problem, just double-checking)

File: client/web/src/enterprise/insights/pages/dashboards/dashboard-view/components/add-insight-modal/query.ts [hunk 7]
```diff
@@ -1,7 +1,67 @@
...
+    AssignableInsight,
+    DashboardInsights,
+    FindInsightsBySearchTermResult,
+    FindInsightsBySearchTermVariables,
+    GroupByField,
+} from '../../../../../../../graphql-operations'
+
+import { DashboardInsight, InsightSuggestion, InsightType } from './types'
+
+export const SYNC_DASHBOARD_INSIGHTS = gql`
```
Comment (valerybugakov) [inline client/web/src/enterprise/insights/pages/dashboards/dashboard-view/components/add-insight-modal/query.ts:19]: Do we need to export that?

File: client/web/src/enterprise/insights/pages/dashboards/dashboard-view/components/add-insight-modal/query.ts [hunk 8]
```diff
@@ -1,7 +1,67 @@
...
+    fragment DashboardInsights on InsightsDashboard {
```
Comment (valerybugakov) [inline client/web/src/enterprise/insights/pages/dashboards/dashboard-view/components/add-insight-modal/query.ts:20]: Is there a binding between this fragment and the query it assumes already fetched these fields? I wanted to find a query where this fragment is used to understand how it relates to the API request sent to the server.

File: client/web/src/enterprise/insights/pages/dashboards/dashboard-view/components/add-insight-modal/query.ts [hunk 9]
```diff
@@ -1,7 +1,67 @@
...
+                }
+            }
+        }
+    }
+`
+
+export function getCachedDashboardInsights(client: ApolloClient<unknown>, dashboardId: string): DashboardInsight[] {
+    const dashboardInsights =
+        client
+            .readFragment<DashboardInsights>({
```
Comment (valerybugakov) [inline client/web/src/enterprise/insights/pages/dashboards/dashboard-view/components/add-insight-modal/query.ts:41]: Interesting stuff! We recently discussed with @philipp-spiess how viable it would be to use [usefragment_experimental](https://www.apollographql.com/docs/react/api/react/hooks-experimental/#usefragment_experimental) to bind React components to their data needs.

File: client/web/src/enterprise/insights/pages/dashboards/dashboard-view/components/add-insight-modal/query.ts [hunk 10]
```diff
@@ -12,23 +72,133 @@ export const GET_ACCESSIBLE_INSIGHTS_LIST = gql`
...
+
+interface UseInsightSuggestionsResult {
+    connection?: Connection<InsightSuggestion>
+    loading: boolean
+    hasNextPage: boolean
+    error?: ApolloError
+    fetchMore: () => void
+}
+
+export function useInsightSuggestions(input: UseInsightSuggestionsInput): UseInsightSuggestionsResult {
```
Comment (valerybugakov) [inline client/web/src/enterprise/insights/pages/dashboards/dashboard-view/components/add-insight-modal/query.ts:98]: Pagination hooks still look scary :)

**Review by valerybugakov — APPROVED** [review]
LGTM! Tested locally and left a couple of optional comments.

File: client/web/src/enterprise/insights/pages/dashboards/dashboard-view/components/add-insight-modal/AddInsightModal.module.scss (hunk 4 above)
Comment (vovakulikov) [reply client/web/src/enterprise/insights/pages/dashboards/dashboard-view/components/add-insight-modal/AddInsightModal.module.scss:6]: Oh yeah (I've had this PR before we added :where to the card styles), I think we can remove it, right?

File: client/web/src/enterprise/insights/pages/dashboards/dashboard-view/components/add-insight-modal/AddInsightModal.tsx (hunk 5 above)
Comment (vovakulikov) [reply client/web/src/enterprise/insights/pages/dashboards/dashboard-view/components/add-insight-modal/AddInsightModal.tsx:86]: I think we can drop this. will do it in a separate PR

File: client/web/src/enterprise/insights/pages/dashboards/dashboard-view/components/add-insight-modal/AddInsightModal.tsx (hunk 6 above)
Comment (vovakulikov) [reply client/web/src/enterprise/insights/pages/dashboards/dashboard-view/components/add-insight-modal/AddInsightModal.tsx:118]: I see what you mean here but I think this is safe because 
- In this particular example there is only one moment when we don't have connection object and fallback on empty (first time list loading phase)
- Even if we call useLayout every time in the MultiCombobox implementation I used mutable ref variable as a store so no additional react call will be triggered

File: client/web/src/enterprise/insights/pages/dashboards/dashboard-view/components/add-insight-modal/query.ts (hunk 7 above)
Comment (vovakulikov) [reply client/web/src/enterprise/insights/pages/dashboards/dashboard-view/components/add-insight-modal/query.ts:19]: Yeah it was needed before, I forgot to remove this, thanks

File: client/web/src/enterprise/insights/pages/dashboards/dashboard-view/components/add-insight-modal/query.ts (hunk 8 above)
Comment (vovakulikov) [reply client/web/src/enterprise/insights/pages/dashboards/dashboard-view/components/add-insight-modal/query.ts:20]: there is no strict binding, I mean we don't use this particular fragment anywhere else but here, it does rely on the cache data, we assume that at the moment when we open this modal we already loaded information about the currently opened dashboard. I added a comment about this and linked the original query on which we rely here.
